File: 백준/Silver/11726. 2×n 타일링/2×n 타일링.cc
#include <iostream>
using namespace std;
using ll = long long;

const ll MAX = 1010;
const ll MOD = 10007;
ll n, dp[MAX];

ll solve(ll cur){
    if(cur < 0) return 0;
    ll& ret = dp[cur];
    if(ret != -1) return ret; ret = 0;
    ret += solve(cur - 1) + solve(cur - 2);

    return ret %= MOD; 
}

int main(){
    ios::sync_with_stdio(0); // fastio
    cin.tie(0), cout.tie(0); // fastio

    cin >> n;
    for(int i = 0;i < MAX;i++) dp[i] = -1;
    dp[0] = 1;

    cout << solve(n);

    return 0;
}

File: 백준/Silver/11726. 2×n 타일링/2×n 타일링.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

dp = [-1] * 1010
mod = 10007
dp[0] = 1

def solve(cur):
    if(cur < 0):
        return 0
    if dp[cur] != -1:
        return dp[cur]
    dp[cur] = 0

    dp[cur] = (solve(cur - 1) + solve(cur - 2)) % mod
    return dp[cur]

n = int(input().rstrip())
print(solve(n))